src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java

```diff
@@ -28,40 +27,46 @@
+
+    private final ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository;
 
     private final Optional<VersionControlService> versionControlService;
 
     private final Optional<ContinuousIntegrationService> continuousIntegrationService;
 
     private final SimpMessageSendingOperations messagingTemplate;
 
-    public ProgrammingSubmissionService(ProgrammingSubmissionRepository programmingSubmissionRepository, ParticipationRepository participationRepository,
-            Optional<VersionControlService> versionControlService, Optional<ContinuousIntegrationService> continuousIntegrationService, ParticipationService participationService,
-            SimpMessageSendingOperations messagingTemplate) {
+    public ProgrammingSubmissionService(ProgrammingSubmissionRepository programmingSubmissionRepository,
+            ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository, Optional<VersionControlService> versionControlService,
+            Optional<ContinuousIntegrationService> continuousIntegrationService, ParticipationService participationService, SimpMessageSendingOperations messagingTemplate,
+            ProgrammingExerciseParticipationService programmingExerciseParticipationService) {
         this.programmingSubmissionRepository = programmingSubmissionRepository;
-        this.participationRepository = participationRepository;
+        this.programmingExerciseStudentParticipationRepository = programmingExerciseStudentParticipationRepository;
         this.versionControlService = versionControlService;
         this.continuousIntegrationService = continuousIntegrationService;
         this.participationService = participationService;
         this.messagingTemplate = messagingTemplate;
+        this.programmingExerciseParticipationService = programmingExerciseParticipationService;
     }
 
-    public void notifyPush(Long participationId, Object requestBody) {
-        Optional<Participation> optionalParticipation = participationRepository.findById(participationId);
-        if (!optionalParticipation.isPresent()) {
-            log.warn("Invalid participation received while notifying about push: " + participationId);
-            return;
-        }
-        Participation participation = optionalParticipation.get();
-        if (participation.getInitializationState() == InitializationState.INACTIVE) {
+    public void notifyPush(Long participationId, Object requestBody) throws IllegalArgumentException {
+        Participation participation = participationService.findOne(participationId);
+        if (!(participation instanceof ProgrammingExerciseParticipation))
+            throw new IllegalArgumentException();
+
+        ProgrammingExerciseParticipation programmingExerciseParticipation = (ProgrammingExerciseParticipation) participation;
+
+        if (participation instanceof ProgrammingExerciseStudentParticipation
```

Comment:
It should be impossible in my opinion that a template or solution build plan is deleted. This is why `resumeExercise` and `triggerBuild` will not be called here for template or solution participations.

Reply:
There would be errors anyway when going through `resumeExercise`, as the logic will try to copy the template participation's build plan.